From 5efa686f444277a7539360ae9ce6e624659ea8f2 Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Tue, 9 Aug 2022 17:34:47 +0800
Subject: [PATCH 4419/4585] [Huawei] exclude bpf_dispatcher_xdp_func from
 mcount list

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187425

----------------

Updating the bpf_dispatcher, bpf_dispatcher_xdp_func, will call
bpf_arch_text_poke and change the fentry instruction.

When ftrace tries to traing bpf_dispatcher_xdp_func, it will find
the fenty is changed in ftrace_verify_code, then ftrace will fail
and throw a warning. If ftrace continues running, it can cause more
problems such as NULL pointer derefrence.

Temporarily, we remove bpf_dispatcher_xdp_func when ftrace init to
prevent this point be traced.

Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
---
 kernel/trace/ftrace.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/kernel/trace/ftrace.c b/kernel/trace/ftrace.c
index 4f40bc2f90a7..2aac1fd8d798 100644
--- a/kernel/trace/ftrace.c
+++ b/kernel/trace/ftrace.c
@@ -6189,6 +6189,8 @@ static int ftrace_process_locs(struct module *mod,
 	unsigned long count;
 	unsigned long *p;
 	unsigned long addr;
+	unsigned long remove_start;
+	unsigned long size = 0;
 	unsigned long flags = 0; /* Shut up gcc */
 	int ret = -ENOMEM;
 
@@ -6228,6 +6230,10 @@ static int ftrace_process_locs(struct module *mod,
 		ftrace_pages->next = start_pg;
 	}
 
+	remove_start = kallsyms_lookup_name("bpf_dispatcher_xdp_func");
+	if (remove_start)
+		kallsyms_lookup_size_offset(remove_start, &size, NULL);
+
 	p = start;
 	pg = start_pg;
 	while (p < end) {
@@ -6238,7 +6244,8 @@ static int ftrace_process_locs(struct module *mod,
 		 * object files to satisfy alignments.
 		 * Skip any NULL pointers.
 		 */
-		if (!addr)
+		if (!addr || (addr >= remove_start &&
+			addr < remove_start + size))
 			continue;
 
 		if (pg->index == pg->size) {
-- 
2.17.1

